[WIP] MATLAB interface

stefphd opened the pull request:
Hi Charlie,

I've just written a first draft of the MATLAB interface. It has also some minor changes in the C/C++ interface. For now it's only a draft, may be unstable at the moment.

Track list:
- [x] Create low-level matlab mex functions `result = uno_optimize(model[, options, callbacks])` and `options = uno_options([preset])`
- [x] Create readme for matlab interface
- [x] Create HS015 problem example
- [x] Use explicit sparsity pattern in matlab low-level interface
- [x] Add argument validation functions in `interfaces\Matlab\Uno_mex_utilities.hpp`
- [x] Add argument validation in`uno_optimize` mex function
- [x] Add high-level (fmicon-like) interface
- [x] Add github action (Windows)
- [x] Add github action (MacOS)

stefphd committed:
Renamed UNO_OPTION_TYPE_INT to UNO_OPTION_TYPE_INTEGER

stefphd committed:
Moved stream callbacks to UserModel.hpp

stefphd committed:
Added Vector(begin, end) constructor

stefphd committed:
Added typed option getters

stefphd committed:
Added Matlab MEX interface

stefphd committed:
Create README for Matlab interface

stefphd committed:
Added Matlab example

stefphd committed:
Fixed missing include

stefphd committed:
Matlab libut not required

stefphd committed:
Added explicit sparsity patterns in matlab interface

stefphd committed:
Updated matlab example to use explicit sparsity

stefphd committed:
Updated matlab README

stefphd committed:
Simplified struct to option conversion

stefphd committed:
Added conversion of matlab string to std::string

stefphd committed:
Accept matlab string in option preset

stefphd committed:
Merge branch 'main' of https://github.com/cvanaret/Uno

stefphd committed:
Removed redundant callbacks in matlab interface

stefphd committed:
Merge branch 'main' of https://github.com/cvanaret/Uno

stefphd committed:
Updated matlab with number model evaluations

stefphd committed:
Updated the matlab example

stefphd committed:
Added function argument validation

stefphd committed:
Reorganized matlab interface sources

stefphd committed:
Link matlab interface to uno default lib

stefphd committed:
Validate user callback outputs during optimization

stefphd committed:
Removed unused functions

stefphd committed:
Update example_hs015.m

stefphd committed:
Added matlab Polak5 example

stefphd committed:
Flip multiplier sign back before eval (data are copied)

stefphd committed:
Update example_polak5.m

stefphd committed:
Renamed validate_matlab_handle to validate_matlab_handle_field

stefphd committed:
Added high-level Matlab interface

stefphd committed:
Fixed missing (:) in uno.m

stefphd committed:
Fixed typo in Matlab readme

stefphd committed:
Added Test Matlab interface workflow

cvanaret commented on interfaces/Matlab/README.md:
```diff
@@ -11,7 +11,7 @@ An example is available in the file [example_hs015.m](example/example_hs015.m).
 ## Affiliation
 
-This Julia interface is developed and maintained by [Stefano Lovato](https://github.com/stefphd) and [Charlie Vanaret](https://github.com/cvanaret).
+This Objective interface is developed and maintained by [Stefano Lovato](https://github.com/stefphd) and [Charlie Vanaret](https://github.com/cvanaret).
```
Still not :smile:

stefphd committed:
Fixed typo in Matlab readme

stefphd committed:
Exclude auto build of Matlab interface

stefphd committed:
Try install Matlab files

stefphd committed:
Added explicit error messages in Uno matlab interface

stefphd committed:
Merge branch 'main' of https://github.com/cvanaret/Uno

stefphd committed:
Fixed unomex_optimize hessian and problem type

stefphd committed:
Clear uno_optimize after calling uno

stefphd committed:
Merge branch 'main' of https://github.com/cvanaret/Uno

stefphd committed:
Merged C API with main changes

stefphd committed:
unomex_* renamed to uno_

stefphd committed:
Added validate_string_field

stefphd committed:
Added missing return

stefphd committed:
Updated Matlab examples with new problem_type

stefphd committed:
Link Matlab libut in unomex-options target

stefphd committed:
Divided C++ classes for Matlab interface into multiple files

cvanaret commented:
@stefphd sorry for the many changes we're merging into main :sweat_smile: Let me know if you need anything.

stefphd committed:
Added Matlab test example on MacOS

stefphd committed:
Fixed platform in github workflow

stefphd committed:
Try adding fortran compiler in macod workflow

stefphd committed:
Fixed wrong slash in include

stefphd committed:
Fixed wrong slash in include (missing)

stefphd committed:
Merge branch 'main' of https://github.com/cvanaret/Uno

stefphd committed:
Updated multiplier sign convention in Materlab interface

stefphd committed:
compile the Unomex source files once

stefphd committed:
Try fix rpath for macos Matlab interface

stefphd committed:
Merge branch 'main' of https://github.com/cvanaret/Uno

stefphd committed:
Merge branch 'main' of https://github.com/cvanaret/Uno

stefphd committed:
Merge branch 'main' of https://github.com/cvanaret/Uno

stefphd committed:
Merge branch 'main' of https://github.com/cvanaret/Uno

stefphd committed:
Update C interface for breaking changes

stefphd committed:
Update CMakeLists.txt for breaking changes

stefphd committed:
Remove old function members (not needed)

stefphd committed:
Update for braking changes in Uno interface

stefphd committed:
Use HessianFnc to mimic Matlab fmincon

Lagrangian Hessian is now specified using options.HessianFnc in the high-level interface, similar to Matlab fmincon. Fallback to LBFGS if no HessianFnc is provided

stefphd committed:
Add minimum equired version

stefphd committed:
Fix typo

stefphd committed:
Update MATLAB documentation

stefphd committed:
Fix wrong matlab install dir

bin for Windows, lib otherwise (mac-os etc)

stefphd commented on CMakeLists.txt:
```diff
@@ -297,6 +297,58 @@ if(pybind11_DIR)
+# optional Matlab bindings #
+############################
+
+find_package(Matlab COMPONENTS MAIN_PROGRAM MEX_COMPILER)
```
Verified up to release R2023b (corresponding to version 23.2). 
Now minimum required version is 23.2, see commit [6313475](https://github.com/cvanaret/Uno/pull/360/commits/631347569b8345c1c9fca9116b926bf08c8fa506).

stefphd committed:
Make sure to print all messages into matlab command window

stefphd committed:
Update matlab-test workflow

stefphd committed:
Fix wrong line separator in Windows configure

stefphd committed:
Try to add pause before exiting to avoid matlab crash

stefphd committed:
Remove pause (not wroking to avoid matlab crash)

stefphd committed:
Use 'cmd' shell for Windows configure

stefphd committed:
Remove options.HessianFnc after use to avoid Uno warning

stefphd committed:
Try to lock mex to prevent matlab crash on cleanup

stefphd committed:
Add link for mexLock() workaround

stefphd committed:
Merge branch 'main' of https://github.com/cvanaret/Uno

stefphd committed:
Add missing result fields

stefphd committed:
Fix typo in error strings

stefphd committed:
Fix wrong line for initial_primal_iterate

stefphd committed:
Handle the option preset

cvanaret committed:
Added missing include

cvanaret committed:
Removed LTO

cvanaret committed:
Fixed linking order of (C)BLAS

stefphd committed:
Merge branch 'main' of https://github.com/stefphd/Uno

stefphd committed:
Fix wrong initial iterate for interior-point
